src/boundary_conditions.jl

```diff
@@ -2,26 +2,87 @@
 ##### The basic idea: boundary condition types and data.
 #####
 
-abstract type BCType end
+"""
+    BCType
+
+Abstract supertype for boundary condition types.
+"""
+abstract type BCType end 
+
+"""
+    Periodic
+
+A type specifying a periodic boundary condition.
+
+A condition may not be specified with a `Periodic` boundary condition.
+"""
 struct Periodic <: BCType end
+
+"""
+    Flux
+
+A type specifying a boundary condition on the flux of a field.
+"""
 struct Flux <: BCType end
+
+"""
+    Gradient
+
+A type specifying a boundary condition on the derivative or gradient of a field. Also
+called a Neumann boundary condition.
+"""
 struct Gradient <: BCType end
+
+"""
+    Value
+
+A type specifying a boundary condition on the value of a field. Also called a Dirchlet
+boundary condition.
+"""
 struct Value <: BCType end
+
+"""
```

Comment:
Hmm, I think the description about how this is different from a no-flux conditions is confusing. In fact, `NoPenetration` resembles a `Value` condition most closely, since it specifies the "value" of a field along a boundary on which the field's points are co-located. 

It may help to be very specific here:

`NoPenetration` can only be applied to `u` along x, `v` along y, or `w` along z.

For all other cases --- fields located at `(Cell, Cell, Cell)`, or `u`, `v`, and `w` in (y, z), (x, z), and (x, y), respectively, either `Value`, `Gradient`, or `Flux` conditions must be used.

Actually, we probably don't expect users to explicitly use this condition, Instead, we will provide `Model` constructors and `SolutionBoundaryConditions` constructors that ensure this behavior for the four possible model configurations we could consider (triply periodic, horizontally-periodic, channel, and closed box).